src/api.c

```diff
@@ -885,7 +889,29 @@ struct nmd_frame *nmd_get_next_frame(struct nmd_ctx *s)
     if (ret < 0)
         return ret_frame(s, NULL);
 
+    if (s->loop) {
+        s->loop = 0;
+
+        av_frame_free(&s->cached_frame);
+        s->last_pushed_frame_ts = AV_NOPTS_VALUE;
+
+        const struct nmdi_opts *o = &s->opts;
+        ret = nmdi_async_seek(s->actx, o->start_time64);
+        if (ret < 0)
+            LOG(s, ERROR, "Failed to seek back to beginning of the file");
+    }
+
     AVFrame *frame = pop_frame(s);
+    if (!frame && s->eof) {
+        /*
+         * If the async module reached EOF and there is no more frame to
+         * return, we need to seek back to the beginning of the file because
+         * nmd_get_next_frame() is expected to loop around when reaching EOF.
+         */
+        s->eof = 0;
+        s->loop = 1;
+    }
```

Comment:
It doesn't work if you remove this block and instead of checking of `s->loop` you check for `s->eof` and the next iteration?

```c
if (s->eof) {
    // seek
}
AVframe *frame = pop_frame(s);
return ret_frame(s, frame);
```

Am I missing something?

Reply:
Checking for eof this way is not enough, you need to actually have reached eof and have no more frame to return.

Reply:
But if the previous call set `s->eof`, it also necessarily returned a `NULL` frame?

Reply:
Ah mmh ok I think I get it, my bad.